deep_copy.go

```diff
@@ -48,72 +92,94 @@ func deepCopyIface(in, out reflect.Value) {
 	case reflect.Ptr:
 		newVal := reflect.New(inElem.Type().Elem())
 		out.Set(newVal)
-		deepCopy(inElem.Elem(), newVal.Elem())
+		d.deepCopy(inElem.Elem(), newVal.Elem())
 		return
 	case reflect.Struct:
 		newVal := reflect.New(inElem.Type())
-		deepCopy(inElem, newVal.Elem())
+		d.deepCopy(inElem, newVal.Elem())
 		out.Set(newVal.Elem())
 		return
 	case reflect.Map:
 		if inElem.IsNil() {
 			return
 		}
 		out.Set(reflect.MakeMapWithSize(inElem.Type(), inElem.Len()))
-		deepCopy(inElem, out.Elem())
+		d.deepCopy(inElem, out.Elem())
 		return
 	case reflect.Slice:
 		if inElem.IsNil() {
 			return
 		}
 		out.Set(reflect.MakeSlice(inElem.Type(), inElem.Len(), inElem.Cap()))
-		deepCopy(inElem, out.Elem())
+		d.deepCopy(inElem, out.Elem())
 		return
 	case reflect.Array:
 		newVal := reflect.New(inElem.Type())
-		deepCopyArray(inElem, newVal.Elem())
+		d.deepCopyArray(inElem, newVal.Elem())
 		out.Set(newVal.Elem())
 		return
 	}
 	// not a pointer, struct or map
 	out.Set(inElem)
 }
-func deepCopyPtr(in, out reflect.Value) {
+func (d *deepCopier) deepCopyPtr(in, out reflect.Value) {
 	if in.IsNil() {
 		return
 	}
+	pKey := ptrKey{ptr: in.Pointer(), typ: in.Type()}
+	if ov, ok := d.ptrMap[pKey]; ok {
+		out.Set(ov)
+		// The deep part of the copying has already been taken care of
+		return
+	}
+
 	inType := in.Type()
 	newVal := reflect.New(inType.Elem())
 	out.Set(newVal)
-	deepCopy(in.Elem(), out.Elem())
+	d.ptrMap[pKey] = out
+	d.deepCopy(in.Elem(), out.Elem())
 }
 
 // deepCopyStruct does a deep-copy of the passed struct-type from in into out.
-func deepCopyStruct(in, out reflect.Value) {
+func (d *deepCopier) deepCopyStruct(in, out reflect.Value) {
 	for i := 0; i < in.NumField(); i++ {
+		ifld := in.Type().Field(i)
 		f := in.Field(i)
 		of := out.Field(i)
-		deepCopy(f, of)
+		if !token.IsExported(ifld.Name) {
+			// The field is not exported, and Go's reflect package
+			// goes to great lengths to consider values from such
+			// fields poisoned. Don't try to bypass it, since we're
+			// dealing with config structs which are assembled from
+			// different sources, not general-purpose values.
+			continue
+		}
+		d.deepCopy(f, of)
 	}
 }
 
 // deepCopySlice allocates a new slice copying values from in, and assigns it
 // to out.
-func deepCopySlice(in, out reflect.Value) {
+func (d *deepCopier) deepCopySlice(in, out reflect.Value) {
 	if in.Kind() != reflect.Slice || out.Kind() != reflect.Slice {
 		panic(fmt.Errorf("unexpected type: in: %s; out: %s", in.Type(), out.Type()))
 	}
 	if in.IsNil() {
 		return
 	}
+
+	if in.Cap() > 0 {
+	}
+
 	if (out.IsNil() || out.Pointer() == in.Pointer()) && out.CanSet() {
```

Comment:
Out of curiosity, when can we have an instance where out is Nil but it can also be set?

Reply:
Essentially, if `out` is `nil`, then it's an empty pointer and we need to allocate new backing storage for it.
`IsNil()` indicates whether we can dereference a value, rather than whether it's settable. In fact, we need to set the return value of `MakeSlice()` to make it useful. If we can't assign to it, and it's not settable we'll probably panic in `deepCopyArray` below.